File: qc-application/src/test/java/com/eager/questioncloud/application/business/payment/question/implement/QuestionPaymentProcessorTest.java
package com.eager.questioncloud.application.business.payment.question.implement;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import com.eager.questioncloud.application.utils.Fixture;
import com.eager.questioncloud.core.domain.coupon.enums.CouponType;
import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.CouponRepository;
import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.UserCouponRepository;
import com.eager.questioncloud.core.domain.coupon.model.Coupon;
import com.eager.questioncloud.core.domain.coupon.model.UserCoupon;
import com.eager.questioncloud.core.domain.payment.infrastructure.repository.QuestionOrderRepository;
import com.eager.questioncloud.core.domain.payment.infrastructure.repository.QuestionPaymentRepository;
import com.eager.questioncloud.core.domain.payment.model.QuestionOrder;
import com.eager.questioncloud.core.domain.payment.model.QuestionPayment;
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentCoupon;
import com.eager.questioncloud.core.domain.point.implement.UserPointManager;
import com.eager.questioncloud.core.domain.point.infrastructure.repository.UserPointRepository;
import com.eager.questioncloud.core.domain.point.model.UserPoint;
import com.eager.questioncloud.core.domain.question.enums.QuestionStatus;
import com.eager.questioncloud.core.domain.question.infrastructure.repository.QuestionRepository;
import com.eager.questioncloud.core.domain.question.model.Question;
import com.eager.questioncloud.core.domain.user.infrastructure.repository.UserRepository;
import com.eager.questioncloud.core.domain.user.model.User;
import com.eager.questioncloud.core.exception.CoreException;
import com.eager.questioncloud.core.exception.Error;
import java.time.LocalDateTime;
import java.util.List;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.SpyBean;
import org.springframework.test.context.ActiveProfiles;

@SpringBootTest
@ActiveProfiles("test")
class QuestionPaymentProcessorTest {
    @Autowired
    QuestionPaymentProcessor questionPaymentProcessor;

    @Autowired
    UserRepository userRepository;

    @Autowired
    QuestionRepository questionRepository;

    @Autowired
    CouponRepository couponRepository;

    @Autowired
    UserCouponRepository userCouponRepository;

    @Autowired
    UserPointRepository userPointRepository;

    @SpyBean
    QuestionOrderRepository questionOrderRepository;

    @SpyBean
    QuestionPaymentRepository questionPaymentRepository;

    @SpyBean
    QuestionPaymentCouponProcessor questionPaymentCouponProcessor;

    @SpyBean
    UserPointManager userPointManager;

    @SpyBean
    QuestionPaymentHistoryGenerator questionPaymentHistoryGenerator;

    @AfterEach
    void tearDown() {
        userRepository.deleteAllInBatch();
        questionRepository.deleteAllInBatch();
        couponRepository.deleteAllInBatch();
        userCouponRepository.deleteAllInBatch();
        questionOrderRepository.deleteAllInBatch();
    }

    @Test
    @DisplayName("문제 결제 처리를 할 수 있다. (쿠폰 O)")
    void processQuestionPayment() {
        // given
        User user = userRepository.save(Fixture.fixtureMonkey.giveMeBuilder(User.class)
            .set("uid", null)
            .sample()
        );

        int beforeUserPoint = 1000000;
        userPointRepository.save(new UserPoint(user.getUid(), beforeUserPoint));

        Coupon coupon = couponRepository.save(
            Fixture.fixtureMonkey.giveMeBuilder(Coupon.class)
                .set("id", null)
                .set("couponType", CouponType.Fixed)
                .set("value", 1000)
                .set("endAt", LocalDateTime.now().plusDays(10))
                .sample()
        );

        UserCoupon userCoupon = userCouponRepository.save(
            Fixture.fixtureMonkey.giveMeBuilder(UserCoupon.class)
                .set("couponId", coupon.getId())
                .set("userId", user.getUid())
                .set("endAt", coupon.getEndAt())
                .set("isUsed", false)
                .sample()
        );

        List<Question> questions = Fixture.fixtureMonkey.giveMeBuilder(Question.class)
            .set("id", null)
            .set("questionContent.questionCategoryId", 25L)
            .set("questionContent.price", 1000)
            .set("questionStatus", QuestionStatus.Available)
            .sampleList(10);

        QuestionPayment questionPayment = QuestionPayment.create(
            user.getUid(),
            QuestionPaymentCoupon.create(userCoupon.getId(), coupon),
            QuestionOrder.createOrder(questions)
        );

        int originalAmount = questionPayment.getAmount();
        int discountedAmount = questionPayment.getQuestionPaymentCoupon().calcDiscount(originalAmount);

        // when
        QuestionPayment paymentResult = questionPaymentProcessor.processQuestionPayment(questionPayment);

        // then
        verify(questionPaymentCouponProcessor, times(1)).applyCoupon(questionPayment);
        verify(userPointManager, times(1)).usePoint(questionPayment.getUserId(), questionPayment.getAmount());
        verify(questionOrderRepository, times(1)).save(questionPayment.getOrder());
        verify(questionPaymentRepository, times(1)).save(questionPayment);
        verify(questionPaymentHistoryGenerator, times(1)).saveQuestionPaymentHistory(paymentResult);

        UserCoupon afterUserCoupon = userCouponRepository.getUserCoupon(userCoupon.getId());
        Assertions.assertThat(afterUserCoupon.getIsUsed()).isTrue();

        Assertions.assertThat(paymentResult.getAmount()).isEqualTo(discountedAmount);

        UserPoint afterUserPoint = userPointRepository.getUserPoint(user.getUid());
        Assertions.assertThat(afterUserPoint.getPoint()).isEqualTo(beforeUserPoint - paymentResult.getAmount());
    }

    @Test
    @DisplayName("문제 결제 처리를 할 수 있다. (쿠폰 X)")
    void processQuestionPaymentNoCoupon() {
        // given
        User user = userRepository.save(Fixture.fixtureMonkey.giveMeBuilder(User.class)
            .set("uid", null)
            .sample()
        );

        int beforeUserPoint = 1000000;
        userPointRepository.save(new UserPoint(user.getUid(), beforeUserPoint));

        List<Question> questions = Fixture.fixtureMonkey.giveMeBuilder(Question.class)
            .set("id", null)
            .set("questionContent.questionCategoryId", 25L)
            .set("questionContent.price", 1000)
            .set("questionStatus", QuestionStatus.Available)
            .sampleList(10);

        QuestionPayment questionPayment = QuestionPayment.create(
            user.getUid(),
            QuestionPaymentCoupon.noCoupon(),
            QuestionOrder.createOrder(questions)
        );

        int originalAmount = questionPayment.getAmount();

        // when
        QuestionPayment paymentResult = questionPaymentProcessor.processQuestionPayment(questionPayment);

        // then
        verify(questionPaymentCouponProcessor, times(1)).applyCoupon(questionPayment);
        verify(userPointManager, times(1)).usePoint(questionPayment.getUserId(), questionPayment.getAmount());
        verify(questionOrderRepository, times(1)).save(questionPayment.getOrder());
        verify(questionPaymentRepository, times(1)).save(questionPayment);
        verify(questionPaymentHistoryGenerator, times(1)).saveQuestionPaymentHistory(paymentResult);

        Assertions.assertThat(paymentResult.getAmount()).isEqualTo(originalAmount);

        UserPoint afterUserPoint = userPointRepository.getUserPoint(user.getUid());
        Assertions.assertThat(afterUserPoint.getPoint()).isEqualTo(beforeUserPoint - paymentResult.getAmount());
    }

    @Test
    @DisplayName("보유 포인트가 부족하면 포인트 부족 예외가 발생한다.")
    void cancelQuestionPaymentWhenNotEnoughUserPoint() {
        // given
        User user = userRepository.save(Fixture.fixtureMonkey.giveMeBuilder(User.class)
            .set("uid", null)
            .sample()
        );

        int beforeUserPoint = 1000;
        userPointRepository.save(new UserPoint(user.getUid(), beforeUserPoint));

        Coupon coupon = couponRepository.save(
            Fixture.fixtureMonkey.giveMeBuilder(Coupon.class)
                .set("id", null)
                .set("couponType", CouponType.Fixed)
                .set("value", 1000)
                .set("endAt", LocalDateTime.now().plusDays(10))
                .sample()
        );

        UserCoupon userCoupon = userCouponRepository.save(
            Fixture.fixtureMonkey.giveMeBuilder(UserCoupon.class)
                .set("couponId", coupon.getId())
                .set("userId", user.getUid())
                .set("endAt", coupon.getEndAt())
                .set("isUsed", false)
                .sample()
        );

        List<Question> questions = Fixture.fixtureMonkey.giveMeBuilder(Question.class)
            .set("id", null)
            .set("questionContent.questionCategoryId", 25L)
            .set("questionContent.price", 1000)
            .set("questionStatus", QuestionStatus.Available)
            .sampleList(10);

        QuestionPayment questionPayment = QuestionPayment.create(
            user.getUid(),
            QuestionPaymentCoupon.create(userCoupon.getId(), coupon),
            QuestionOrder.createOrder(questions)
        );

        // when then
        Assertions.assertThatThrownBy(() -> questionPaymentProcessor.processQuestionPayment(questionPayment))
            .isInstanceOf(CoreException.class)
            .hasFieldOrPropertyWithValue("error", Error.NOT_ENOUGH_POINT);
    }

    @Test
    @DisplayName("결제 도중 예외가 발생하면 쿠폰, 포인트가 롤백 처리 된다.")
    void rollbackWhenOccurException() {
        // given
        User user = userRepository.save(Fixture.fixtureMonkey.giveMeBuilder(User.class)
            .set("uid", null)
            .sample()
        );

        int beforeUserPoint = 1000000;
        userPointRepository.save(new UserPoint(user.getUid(), beforeUserPoint));

        Coupon coupon = couponRepository.save(
            Fixture.fixtureMonkey.giveMeBuilder(Coupon.class)
                .set("id", null)
                .set("couponType", CouponType.Fixed)
                .set("value", 1000)
                .set("endAt", LocalDateTime.now().plusDays(10))
                .sample()
        );

        UserCoupon userCoupon = userCouponRepository.save(
            Fixture.fixtureMonkey.giveMeBuilder(UserCoupon.class)
                .set("couponId", coupon.getId())
                .set("userId", user.getUid())
                .set("endAt", coupon.getEndAt())
                .set("isUsed", false)
                .sample()
        );

        List<Question> questions = Fixture.fixtureMonkey.giveMeBuilder(Question.class)
            .set("id", null)
            .set("questionContent.questionCategoryId", 25L)
            .set("questionContent.price", 1000)
            .set("questionStatus", QuestionStatus.Available)
            .sampleList(10);

        QuestionPayment questionPayment = QuestionPayment.create(
            user.getUid(),
            QuestionPaymentCoupon.create(userCoupon.getId(), coupon),
            QuestionOrder.createOrder(questions)
        );

        Mockito.doThrow(new CoreException(Error.PAYMENT_ERROR)).when(questionPaymentRepository).save(any());

        // when then
        Assertions.assertThatThrownBy(() -> questionPaymentProcessor.processQuestionPayment(questionPayment))
            .isInstanceOf(CoreException.class);

        UserCoupon afterUserCoupon = userCouponRepository.getUserCoupon(userCoupon.getId());
        Assertions.assertThat(afterUserCoupon.getIsUsed()).isFalse();

        UserPoint afterUserPoint = userPointRepository.getUserPoint(user.getUid());
        Assertions.assertThat(afterUserPoint.getPoint()).isEqualTo(beforeUserPoint);
    }
}

File: qc-application/src/test/java/com/eager/questioncloud/application/business/payment/question/implement/QuestionPaymentProcessorTest.kt
package com.eager.questioncloud.application.business.payment.question.implement

import com.eager.questioncloud.application.utils.Fixture
import com.eager.questioncloud.core.domain.coupon.enums.CouponType
import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.CouponRepository
import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.UserCouponRepository
import com.eager.questioncloud.core.domain.coupon.model.Coupon
import com.eager.questioncloud.core.domain.coupon.model.UserCoupon
import com.eager.questioncloud.core.domain.payment.infrastructure.repository.QuestionOrderRepository
import com.eager.questioncloud.core.domain.payment.infrastructure.repository.QuestionPaymentRepository
import com.eager.questioncloud.core.domain.payment.model.QuestionOrder.Companion.createOrder
import com.eager.questioncloud.core.domain.payment.model.QuestionPayment.Companion.create
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentCoupon.Companion.create
import com.eager.questioncloud.core.domain.point.implement.UserPointManager
import com.eager.questioncloud.core.domain.point.infrastructure.repository.UserPointRepository
import com.eager.questioncloud.core.domain.point.model.UserPoint
import com.eager.questioncloud.core.domain.question.enums.QuestionStatus
import com.eager.questioncloud.core.domain.question.infrastructure.repository.QuestionRepository
import com.eager.questioncloud.core.domain.question.model.Question
import com.eager.questioncloud.core.domain.question.model.QuestionContent
import com.eager.questioncloud.core.domain.user.infrastructure.repository.UserRepository
import com.eager.questioncloud.core.domain.user.model.User
import com.eager.questioncloud.core.exception.CoreException
import com.eager.questioncloud.core.exception.Error
import com.navercorp.fixturemonkey.kotlin.giveMeKotlinBuilder
import com.navercorp.fixturemonkey.kotlin.into
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.SpyBean
import org.springframework.test.context.ActiveProfiles
import java.time.LocalDateTime

@SpringBootTest
@ActiveProfiles("test")
internal class QuestionPaymentProcessorTest {
    @Autowired
    var questionPaymentProcessor: QuestionPaymentProcessor? = null

    @Autowired
    var userRepository: UserRepository? = null

    @Autowired
    var questionRepository: QuestionRepository? = null

    @Autowired
    var couponRepository: CouponRepository? = null

    @Autowired
    var userCouponRepository: UserCouponRepository? = null

    @Autowired
    var userPointRepository: UserPointRepository? = null

    @SpyBean
    var questionOrderRepository: QuestionOrderRepository? = null

    @SpyBean
    var questionPaymentRepository: QuestionPaymentRepository? = null

    @SpyBean
    var questionPaymentCouponProcessor: QuestionPaymentCouponProcessor? = null

    @SpyBean
    var userPointManager: UserPointManager? = null

    @SpyBean
    var questionPaymentHistoryRegister: QuestionPaymentHistoryRegister? = null

    @AfterEach
    fun tearDown() {
        userRepository!!.deleteAllInBatch()
        questionRepository!!.deleteAllInBatch()
        couponRepository!!.deleteAllInBatch()
        userCouponRepository!!.deleteAllInBatch()
        questionOrderRepository!!.deleteAllInBatch()
    }

    @Test
    @DisplayName("문제 결제 처리를 할 수 있다. (쿠폰 O)")
    fun payment() {
        // given
        val user = userRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<User>()
                .set(User::uid, null)
                .build()
                .sample()
        )

        val beforeUserPoint = 1000000
        userPointRepository!!.save(UserPoint(user.uid!!, beforeUserPoint))

        val coupon = couponRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<Coupon>()
                .set(Coupon::id, null)
                .set(Coupon::value, 1000)
                .set(Coupon::couponType, CouponType.Fixed)
                .set(Coupon::endAt, LocalDateTime.now().plusDays(10))
                .sample()
        )

        val userCoupon = userCouponRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<UserCoupon>()
                .set(UserCoupon::couponId, coupon.id)
                .set(UserCoupon::userId, user.uid)
                .set(UserCoupon::endAt, coupon.endAt)
                .set(UserCoupon::isUsed, false)
                .sample()
        )

        val questions = Fixture.fixtureMonkey.giveMeKotlinBuilder<Question>()
            .set(Question::id, null)
            .set(Question::questionContent into QuestionContent::questionCategoryId, 25L)
            .set(Question::questionContent into QuestionContent::price, 1000)
            .set(Question::questionStatus, QuestionStatus.Available)
            .sampleList(10)
            .stream()
            .map { question ->
                questionRepository!!.save(question)
            }
            .toList()

        val questionPayment = create(
            user.uid!!,
            create(userCoupon.id!!, coupon),
            createOrder(questions)
        )

        val originalAmount = questionPayment.amount
        val discountedAmount = questionPayment.questionPaymentCoupon!!.calcDiscount(originalAmount)

        // when
        val paymentResult = questionPaymentProcessor!!.payment(questionPayment)

        // then
        Mockito.verify(questionPaymentCouponProcessor, Mockito.times(1))!!.applyCoupon(questionPayment)
        Mockito.verify(userPointManager, Mockito.times(1))!!.usePoint(questionPayment.userId, questionPayment.amount)
        Mockito.verify(questionPaymentRepository, Mockito.times(1))!!.save(questionPayment)

        val afterUserCoupon = userCouponRepository!!.getUserCoupon(userCoupon.id!!)
        Assertions.assertThat(afterUserCoupon.isUsed).isTrue()

        Assertions.assertThat(questionPayment.amount).isEqualTo(discountedAmount)

        val afterUserPoint = userPointRepository!!.getUserPoint(user.uid!!)
        Assertions.assertThat(afterUserPoint.point).isEqualTo(beforeUserPoint - questionPayment.amount)
    }

    @Test
    @DisplayName("문제 결제 처리를 할 수 있다. (쿠폰 X)")
    fun paymentNoCoupon() {
        // given
        val user = userRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<User>()
                .set(User::uid, null)
                .build()
                .sample()
        )

        val beforeUserPoint = 1000000
        userPointRepository!!.save(UserPoint(user.uid!!, beforeUserPoint))

        val questions = Fixture.fixtureMonkey.giveMeKotlinBuilder<Question>()
            .set(Question::id, null)
            .set(Question::questionContent into QuestionContent::questionCategoryId, 25L)
            .set(Question::questionContent into QuestionContent::price, 1000)
            .set(Question::questionStatus, QuestionStatus.Available)
            .sampleList(10)
            .stream()
            .map { question ->
                questionRepository!!.save(question)
            }
            .toList()

        val questionPayment = create(
            user.uid!!,
            null,
            createOrder(questions)
        )

        val originalAmount = questionPayment.amount

        // when
        val paymentResult = questionPaymentProcessor!!.payment(questionPayment)

        // then
        Mockito.verify(questionPaymentCouponProcessor, Mockito.times(1))!!.applyCoupon(questionPayment)
        Mockito.verify(userPointManager, Mockito.times(1))!!.usePoint(questionPayment.userId, questionPayment.amount)
        Mockito.verify(questionPaymentRepository, Mockito.times(1))!!.save(questionPayment)

        Assertions.assertThat(questionPayment.amount).isEqualTo(originalAmount)

        val afterUserPoint = userPointRepository!!.getUserPoint(user.uid!!)
        Assertions.assertThat(afterUserPoint.point).isEqualTo(beforeUserPoint - questionPayment.amount)
    }

    @Test
    @DisplayName("보유 포인트가 부족하면 포인트 부족 예외가 발생한다.")
    fun cancelQuestionPaymentWhenNotEnoughUserPoint() {
        // given
        val user = userRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<User>()
                .set(User::uid, null)
                .build()
                .sample()
        )

        val beforeUserPoint = 1000
        userPointRepository!!.save(UserPoint(user.uid!!, beforeUserPoint))

        val coupon = couponRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<Coupon>()
                .set(Coupon::id, null)
                .set(Coupon::value, 1000)
                .set(Coupon::couponType, CouponType.Fixed)
                .set(Coupon::endAt, LocalDateTime.now().plusDays(10))
                .sample()
        )

        val userCoupon = userCouponRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<UserCoupon>()
                .set(UserCoupon::couponId, coupon.id)
                .set(UserCoupon::userId, user.uid)
                .set(UserCoupon::endAt, coupon.endAt)
                .set(UserCoupon::isUsed, false)
                .sample()
        )

        val questions = Fixture.fixtureMonkey.giveMeKotlinBuilder<Question>()
            .set(Question::id, null)
            .set(Question::questionContent into QuestionContent::questionCategoryId, 25L)
            .set(Question::questionContent into QuestionContent::price, 1000)
            .set(Question::questionStatus, QuestionStatus.Available)
            .sampleList(10)
            .stream()
            .map { question ->
                questionRepository!!.save(question)
            }
            .toList()

        val questionPayment = create(
            user.uid!!,
            create(userCoupon.id!!, coupon),
            createOrder(questions)
        )

        // when then
        Assertions.assertThatThrownBy { questionPaymentProcessor!!.payment(questionPayment) }
            .isInstanceOf(CoreException::class.java)
            .hasFieldOrPropertyWithValue("error", Error.NOT_ENOUGH_POINT)
    }

    @Test
    @DisplayName("결제 도중 예외가 발생하면 쿠폰, 포인트가 롤백 처리 된다.")
    fun rollbackWhenOccurException() {
        // given
        val user = userRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<User>()
                .set(User::uid, null)
                .build()
                .sample()
        )

        val beforeUserPoint = 1000000
        userPointRepository!!.save(UserPoint(user.uid!!, beforeUserPoint))

        val coupon = couponRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<Coupon>()
                .set(Coupon::id, null)
                .set(Coupon::value, 1000)
                .set(Coupon::couponType, CouponType.Fixed)
                .set(Coupon::endAt, LocalDateTime.now().plusDays(10))
                .sample()
        )

        val userCoupon = userCouponRepository!!.save(
            Fixture.fixtureMonkey.giveMeKotlinBuilder<UserCoupon>()
                .set(UserCoupon::couponId, coupon.id)
                .set(UserCoupon::userId, user.uid)
                .set(UserCoupon::endAt, coupon.endAt)
                .set(UserCoupon::isUsed, false)
                .sample()
        )

        val questions = Fixture.fixtureMonkey.giveMeKotlinBuilder<Question>()
            .set(Question::id, null)
            .set(Question::questionContent into QuestionContent::questionCategoryId, 25L)
            .set(Question::questionContent into QuestionContent::price, 1000)
            .set(Question::questionStatus, QuestionStatus.Available)
            .sampleList(10)
            .stream()
            .map { question ->
                questionRepository!!.save(question)
            }
            .toList()

        val questionPayment = create(
            user.uid!!,
            create(userCoupon.id!!, coupon),
            createOrder(questions)
        )

        Mockito.doThrow(CoreException(Error.PAYMENT_ERROR)).`when`(questionPaymentRepository)!!
            .save(any())

        // when then
        Assertions.assertThatThrownBy { questionPaymentProcessor!!.payment(questionPayment) }
            .isInstanceOf(CoreException::class.java)

        val afterUserCoupon = userCouponRepository!!.getUserCoupon(userCoupon.id!!)
        Assertions.assertThat(afterUserCoupon.isUsed).isFalse()

        val afterUserPoint = userPointRepository!!.getUserPoint(user.uid!!)
        Assertions.assertThat(afterUserPoint.point).isEqualTo(beforeUserPoint)
    }
}